File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.py
def solution(number, k):
    max_len = len(number) - k
    result = 0
    stack = []
    
    for i in number:
        while len(stack) > 0 and k > 0 and stack[-1] < i:
            stack.pop()
            k -= 1
        stack.append(i)
                
    
    return "".join(stack[:max_len])
    

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.java
import java.util.*;

class Solution {
    public String solution(String number, int k) {
        Deque<Character> stack = new ArrayDeque<>();
        int removeCnt = 0;

        for (char c : number.toCharArray()) {
            while (!stack.isEmpty() && stack.peekLast() < c && removeCnt < k) {
                stack.pollLast();
                removeCnt++;
            }
            stack.addLast(c);
        }

        while (removeCnt < k) {
            stack.pollLast();
            removeCnt++;
        }

        StringBuilder sb = new StringBuilder();
        for (char c : stack) sb.append(c);

        return sb.toString();
    }
}
